src/ast/semantic_analyser.cpp

```diff
@@ -1396,6 +1468,29 @@ void SemanticAnalyser::visit(Binop &binop)
   }
 
   binop.type = CreateInteger(64, is_signed);
+
+  auto addr_lhs = binop.left->type.GetAS();
+  auto addr_rhs = binop.right->type.GetAS();
+
+  // if lhs or rhs has different addrspace (not none), then set the
+  // addrspace to none. This preserves the behaviour for x86.
+  if (addr_lhs != addr_rhs && addr_lhs != AddrSpace::none &&
+      addr_rhs != AddrSpace::none)
+  {
+    if (is_final_pass())
+      LOG(WARNING) << "Addrspace mismatch";
```

Comment:
```suggestion
      LOG(WARNING, binop.loc, out_) << "Addrspace mismatch";
```

Reply:
Done.

Reply:
I just remembered previously I discussed this topic: https://github.com/iovisor/bpftrace/pull/1141#discussion_r418320486. In short, pointer type would help to eliminate outer uptr() but it could break the compatibility... https://github.com/iovisor/bpftrace/issues/1507#issuecomment-696369158. I think the current behavior is OK for now.